covid/utils.py

```diff
@@ -0,0 +1,63 @@
+from pathlib import Path
+import time
+import json
+import os
+
+from langdetect import detect as detect_language
+from wasabi import msg
+import pandas as pd
+
+
+SCRIPT_PATH = os.path.abspath(os.path.dirname(__file__))
+ARTICLES_DIR = os.path.join(SCRIPT_PATH, '../data/raw/CORD-19-research-challenge/')
+
+
+def get_language(text):
+    try:
+        language = detect_language(text)
+    except:
+        language = "unknown"
+    return language
+
+def yield_data(articles_dir):
+    meta_path = articles_dir + 'metadata.csv'
+    metadata = pd.read_csv(meta_path, low_memory=False)
+
+    metadata.drop_duplicates(subset=['pmcid'], inplace=True)
+    metadata.drop_duplicates(subset=['sha'], inplace=True)
```

Comment:
i think the issue with this is that it'll see None's as duplicates - of which there are quite a few.
https://stackoverflow.com/questions/23512339/drop-duplicates-while-preserving-nans-in-pandas
doesn't look like there is a particularly nice work around though

Reply:
how about dropping duplicates, concatenating the results and dropping duplicate cord_uid.

Reply:
hm I'm not sure what cord_uid is, so not sure about that. I'm actually not sure if this is the best place to drop duplicates.
The reason being is what if in the metadata you had:
`metadata = pd.DataFrame([{'pmcid':PMC123,  'sha': None}, {'pmcid':PMC123, 'sha': 'efg'}])`
then deduplicated you get
`metadata = pd.DataFrame([{'pmcid':PMC123,  'sha': None'}])`
then when you were querying the papers there was one with a paper id of 'efg' but not one with 'PMC123', then you will have got rid of the relevant metadata.

Reply:
I did a sha2pmcid = {} dict for any sha's which also had pmcids, and then used this to check if any of my loaded papers had an associated pmcid which had already been saved

Reply:
`cord_uid` is a unique identifier for each row.

the problem you are raising is spot on. if we duplicate the metadata and then remove duplicates twice, one for pmcid and one for sha and then concatenate and remove duplicates by cord_uid, i think we are ok.
```
    sha_metadata = metadata.copy()
    sha_metadata = sha_metadata.drop_duplicates(subset=['sha'])
    pmc_metadata = metadata.drop_duplicates(subset=['pmcid'])
    
    metadata = pd.concat([sha_metadata, pmc_metadata])
    metadata.drop_duplicates(subset=['cord_uid'])
```

let me try and see what happens

Reply:
wait the example you are suggesting is exactly the point, isn't it? i want to remove the `sha` that also have a pmcid

Reply:
my point is what if there is a json which has paper id in sha form which you can no longer find in the metadata because it has been deduplicated, but also there was no json for the pmcid. i.e. in my example above - the scenario where PMC123.json doesnt exist. does that make sense?

Reply:
xixi, quite complicated, ok let me think about it and find a good way forward :p